src/core/ucc_lib.c

```diff
@@ -0,0 +1,220 @@
+        if (UCS_OK != ucc_component_is_loaded(ccms->names[i])) {
+            ucc_error("Required TL: \"%s\" (ucc_ccm_%s.so) is not available\n",
+                      ccms->names[i], ccms->names[i]);
+            return UCC_ERR_NO_MESSAGE;
+        }
+    }
+    return UCC_OK;
+}
+
+
+static ucc_status_t ucc_lib_init_filtered(const ucc_lib_params_t *params,
+                                          const ucc_lib_config_t *config,
+                                          ucc_lib_info_t *lib)
+{
+    int n_ccms = ucc_lib_data.n_ccms_loaded;
+    ucc_ccm_iface_t *ccm_iface;
+    ucc_ccm_lib_t   *ccm_lib;
+    ucc_ccm_lib_config_t *ccm_config;
+    ucc_status_t status;
+    int i, need_ccm_name_check;
+
+    lib->libs = (ucc_ccm_lib_t**)malloc(sizeof(ucc_ccm_lib_t*)*n_ccms);
+    if (!lib->libs) {
+        status = UCC_ERR_NO_MEMORY;
+        goto error;
+    }
+    assert(config->ccms.count >= 1);
+    need_ccm_name_check = (0 != strcmp(config->ccms.names[0], "all"));
+    lib->n_libs_opened = 0;
+    for (i=0; i<n_ccms; i++) {
+        ccm_iface = ucc_lib_data.ccm_ifaces[i];
+        if (need_ccm_name_check &&
+            -1 == ucs_config_names_search(config->ccms, ccm_iface->name)) {
+            continue;
+        }
+        CHECK_LIB_CONFIG_CAP(thread_mode,  THREAD_MODE);
+        CHECK_LIB_CONFIG_CAP(coll_types,   COLL_TYPES);
+        ccm_config  = malloc(ccm_iface->ccm_lib_config.size);
+        ucs_config_parser_fill_opts(ccm_config, ccm_iface->ccm_lib_config.table,
+                                    config->full_prefix, ccm_iface->ccm_lib_config.prefix, 0);
+        status = ccm_iface->init(params, config, ccm_config, &ccm_lib);
+
+        if (UCS_OK != status) {
+            ucs_config_parser_release_opts(ccm_config, ccm_iface->ccm_lib_config.table);
+            ucc_error("lib_init failed for component: %s\n", ccm_iface->name);
+            goto error;
+        }
+        ccm_lib->log_component = ccm_config->log_component;
+        snprintf(ccm_lib->log_component.name, strlen(ccm_iface->ccm_lib_config.prefix),
+                 "%s", ccm_iface->ccm_lib_config.prefix);
+        ccm_lib->priority = (-1 == ccm_config->priority) ?
+            ccm_iface->priority : ccm_config->priority;
+        ucs_config_parser_release_opts(ccm_config, ccm_iface->ccm_lib_config.table);
+        lib->libs[lib->n_libs_opened++] = ccm_lib;
+        ucc_info("lib_prefix \"%s\": initialized component \"%s\" priority %d\n",
+                 config->full_prefix, ccm_iface->name, ccm_lib->priority);
+    }
+    return UCS_OK;
+
+error:
+    if (lib->libs) free(lib->libs);
+    return status;
+}
+
+ucc_status_t ucc_lib_init(const ucc_lib_params_t *params,
+                          const ucc_lib_config_h config,
+                          ucc_lib_h *ucc_lib)
+{
+    ucs_status_t status;
+    ucc_lib_info_t *lib;
+
+    if (ucc_lib_data.n_ccms_loaded == 0) {
+        return UCC_ERR_NO_MESSAGE;
+    }
+
+    lib = malloc(sizeof(*lib));
```

Comment:
we'd probably want to do something like "memtrack.h" for UCX... for now maybe doing `#define ucs_malloc malloc` is enough, but if we're exploring UCS - memory tracking is high on my list. :)